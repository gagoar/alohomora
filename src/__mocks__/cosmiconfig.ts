
export const search = jest.fn();
export const cosmiconfig = () => ({ search });