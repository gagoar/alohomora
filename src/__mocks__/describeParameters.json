[
  {
    "Name": "/my-company/my-app/all/GRAPHQL_ENDPOINT",
    "Type": "SecureString",
    "KeyId": "alias/aws/ssm",
    "LastModifiedDate": "2018-11-29T21:26:47.824Z",
    "LastModifiedUser": "arn:aws:iam::687711712713:user/serious",
    "Version": 1,
    "Tier": "Standard",
    "Policies": [],
    "DataType": "text"
  },
  {
    "Name": "/my-company/my-app/all/PAGERDUTY_E2E_STEP_FUNCTION_INTEGRATION_TOKEN",
    "Type": "SecureString",
    "KeyId": "alias/aws/ssm",
    "LastModifiedDate": "2020-04-13T22:49:25.960Z",
    "LastModifiedUser": "arn:aws:iam::687711712713:user/rubeus",
    "Version": 1,
    "Tier": "Standard",
    "Policies": [],
    "DataType": "text"
  },
  {
    "Name": "/my-company/my-app/all/PAGERDUTY_E2E_TEST_FAILURES_INTEGRATION_TOKEN",
    "Type": "SecureString",
    "KeyId": "alias/aws/ssm",
    "LastModifiedDate": "2020-04-21T21:32:08.880Z",
    "LastModifiedUser": "arn:aws:iam::687711712713:user/rubeus",
    "Version": 1,
    "Tier": "Standard",
    "Policies": [],
    "DataType": "text"
  },
  {
    "Name": "/my-company/my-app/all/PHABRICATOR_CONDUIT_TOKEN",
    "Type": "SecureString",
    "KeyId": "alias/aws/ssm",
    "LastModifiedDate": "2018-11-23T19:25:09.741Z",
    "LastModifiedUser": "arn:aws:iam::687711712713:user/serious",
    "Version": 1,
    "Tier": "Standard",
    "Policies": [],
    "DataType": "text"
  },
  {
    "Name": "/my-company/my-app/all/PRISMA_ENDPOINT",
    "Type": "SecureString",
    "KeyId": "alias/aws/ssm",
    "LastModifiedDate": "2018-11-23T19:26:02.454Z",
    "LastModifiedUser": "arn:aws:iam::687711712713:user/serious",
    "Version": 1,
    "Tier": "Standard",
    "Policies": [],
    "DataType": "text"
  },
  {
    "Name": "/my-company/my-app/all/PRISMA_MANAGEMENT_API_SECRET",
    "Type": "SecureString",
    "KeyId": "alias/aws/ssm",
    "LastModifiedDate": "2019-01-22T20:46:29.453Z",
    "LastModifiedUser": "arn:aws:iam::687711712713:user/serious",
    "Version": 4,
    "Tier": "Standard",
    "Policies": [],
    "DataType": "text"
  },
  {
    "Name": "/my-company/my-app/all/SENTRY_DSN",
    "Type": "SecureString",
    "KeyId": "alias/aws/ssm",
    "LastModifiedDate": "2018-11-23T19:27:19.165Z",
    "LastModifiedUser": "arn:aws:iam::687711712713:user/serious",
    "Version": 1,
    "Tier": "Standard",
    "Policies": [],
    "DataType": "text"
  },
  {
    "Name": "/my-company/my-app/all/SENTRY_ENVIRONMENT",
    "Type": "SecureString",
    "KeyId": "alias/aws/ssm",
    "LastModifiedDate": "2018-11-25T21:11:45.969Z",
    "LastModifiedUser": "arn:aws:iam::687711712713:user/serious",
    "Version": 1,
    "Tier": "Standard",
    "Policies": [],
    "DataType": "text"
  },
  {
    "Name": "/my-company/my-app/development/DEPLOY_SERVICE_STEP_FUNCTION_NAME",
    "Type": "SecureString",
    "KeyId": "alias/aws/ssm",
    "LastModifiedDate": "2019-07-12T00:21:37.435Z",
    "LastModifiedUser": "arn:aws:iam::687711712713:user/severus",
    "Version": 1,
    "Tier": "Standard",
    "Policies": [],
    "DataType": "text"
  },
  {
    "Name": "/my-company/my-app/development/DEPLOY_SERVICE_STEP_FUNCTION_WRAPPER_NAME",
    "Type": "SecureString",
    "KeyId": "alias/aws/ssm",
    "LastModifiedDate": "2019-10-03T19:44:53.354Z",
    "LastModifiedUser": "arn:aws:iam::687711712713:user/severus",
    "Version": 1,
    "Tier": "Standard",
    "Policies": [],
    "DataType": "text"
  },
  {
    "Name": "/my-company/my-app/production/GRAPHQL_ENDPOINT",
    "Type": "SecureString",
    "KeyId": "alias/aws/ssm",
    "LastModifiedDate": "2018-11-29T21:26:47.824Z",
    "LastModifiedUser": "arn:aws:iam::687711712713:user/serious",
    "Version": 1,
    "Tier": "Standard",
    "Policies": [],
    "DataType": "text"
  }
]
