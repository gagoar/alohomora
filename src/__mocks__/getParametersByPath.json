[
  {
    "Name": "/my-company/my-app/all/BEAUXBATONS",
    "Type": "String",
    "Value": "France",
    "Version": 2,
    "LastModifiedDate": "2020-04-28T23:38:01.257Z",
    "ARN": "arn:aws:ssm:us-west-2:687711712713:parameter/my-company/my-app/all/BEAUXBATONS",
    "DataType": "text"
  },
  {
    "Name": "/my-company/my-app/all/DURMSTRANG",
    "Type": "String",
    "Value": "Germany",
    "Version": 1,
    "LastModifiedDate": "2020-04-26T02:05:08.789Z",
    "ARN": "arn:aws:ssm:us-west-2:687711712713:parameter/my-company/my-app/all/DURMSTRANG",
    "DataType": "text"
  },
  {
    "Name": "/my-company/my-app/staging/DURMSTRANG",
    "Type": "String",
    "Value": "Norway ",
    "Version": 1,
    "LastModifiedDate": "2020-04-26T02:05:08.789Z",
    "ARN": "arn:aws:ssm:us-west-2:687711712713:parameter/my-company/my-app/staging/DURMSTRANG",
    "DataType": "text"
  },
  {
    "Name": "/my-company/my-app/all/HOGWARTS",
    "Type": "String",
    "Value": "Grand Britain",
    "Version": 1,
    "LastModifiedDate": "2020-04-26T01:58:07.989Z",
    "ARN": "arn:aws:ssm:us-west-2:687711712713:parameter/my-company/my-app/all/HOGWARTS",
    "DataType": "text"
  },
  {
    "Name": "/my-company/my-app/production/HOGWARTS",
    "Type": "String",
    "Value": "somewhere in the magical world",
    "Version": 3,
    "LastModifiedDate": "2020-05-06T18:10:03.788Z",
    "ARN": "arn:aws:ssm:us-west-2:687711712713:parameter/my-company/my-app/production/HOGWARTS",
    "DataType": "text"
  }
]
