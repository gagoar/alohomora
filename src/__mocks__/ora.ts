export const stopAndPersist = jest.fn();
export const fail = jest.fn();
export default () =>
  ({
    start: jest.fn(() => ({
      stopAndPersist,
      fail,
    }))
  })