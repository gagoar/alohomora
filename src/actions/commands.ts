export { command as listCommand } from './listParameters';
export { command as getCommand } from './getParameter';
export { command as setCommand } from './setParameter';
export { command as exportCommand } from './exportAsTemplate';
export { command as deleteCommand } from './deleteParameter';