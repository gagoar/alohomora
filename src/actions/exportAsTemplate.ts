import SSM from 'aws-sdk/clients/ssm';
import groupBy from 'lodash.groupby';

import { Actions } from '../types';
import { REGION, API_VERSION, MAX_RESULTS_FOR_PATH, Environment, Template } from '../utils/constants';
import { normalizeSecretKey } from '../utils/normalizeSecrets';
import { isValidTemplate } from '../utils/guards';
import { getGlobalOptions, Command } from '../utils/getGlobalOptions';
import { setAWSCredentials } from '../utils/setAWSCredentials';
interface Input extends Actions { templateName?: Template, custom?: string };

const getParametersByPath = async (params: SSM.GetParametersByPathRequest, region: string): Promise<SSM.ParameterList> => {

  const ssm = new SSM({ apiVersion: API_VERSION, region });
  const { Parameters: parameters = [], NextToken } = await ssm.getParametersByPath(params).promise();

  if (NextToken) {
    const moreParameters = await getParametersByPath({ ...params, NextToken }, region);
    return [...parameters, ...moreParameters];
  } else {
    return parameters;
  }
}

const getKeys = async ({ prefix, region }: { prefix: string, region: string }) => {
  const parameters = await getParametersByPath({
    Path: `/${prefix}`,
    MaxResults: MAX_RESULTS_FOR_PATH,
    Recursive: true,
    WithDecryption: true,
  }, region);

  const secrets = parameters.map(({ Name = '', Value: value = '' }) => {

    const { name, environment } = normalizeSecretKey(Name, prefix);
    return {
      name,
      environment,
      value,
    }
  });

  return groupBy(secrets, ({ environment }: { environment: string }) => environment);
}
const templateFunctions = {
  [Template.shell]: (secrets: Record<string, string>): string => {
    return Object.keys(secrets).map(key => `export ${key}='${secrets[key]}'`).join('\n');
  },
  [Template.json]: (secrets: Record<string, string>): string => {
    return JSON.stringify(secrets, null, 2);
  },
}

type Keys = { name: string; environment: string; value: string; };

const normalizeGroup = (keys: Keys[]): Record<string, string> => {
  return keys.reduce(((memo, secret) => {
    return Object.keys(memo).length ? { ...memo, [secret.name]: secret.value } : { [secret.name]: secret.value };
  }), {} as Record<string, string>);
};

export const exportAsTemplate = async ({ prefix, environment = Environment.all, region = REGION, templateName = Template.shell }: Input): Promise<string> => {

  try {
    const groupedSecrets = await getKeys({
      prefix,
      region,
    });

    let secrets: Record<string, string>;
    const baseSecrets = normalizeGroup(groupedSecrets[Environment.all] || []);

    if (environment !== Environment.all && environment in groupedSecrets) {
      const envSecrets = normalizeGroup(groupedSecrets[environment]);

      secrets = { ...baseSecrets, ...envSecrets };
    } else {
      secrets = baseSecrets;
    }

    return templateFunctions[templateName](secrets);

  } catch (e) {
    console.error('We found an error trying to retrieve secrets', e);
    throw e;
  }
};

export const command = async (templateName: string = Template.shell, command: Command): Promise<void> => {

  if (isValidTemplate(templateName)) {
    const { params, credentials } = await getGlobalOptions(command);

    setAWSCredentials(credentials);
    const response = await exportAsTemplate({ ...params, templateName });
    console.log(response);
  } else {
    console.error(`please provide a valid templateName (${Template})`);
    process.exit(1)
  }

}
