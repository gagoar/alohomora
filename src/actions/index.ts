export { listParameters } from './listParameters';
export { getParameter } from './getParameter';
export { setParameter } from './setParameter';
export { deleteParameter } from './deleteParameter';
export { exportAsTemplate } from './exportAsTemplate';