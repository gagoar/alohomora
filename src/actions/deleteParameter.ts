import SSM from 'aws-sdk/clients/ssm';
import { AWSError } from 'aws-sdk/lib/core';
import ora from 'ora';

import { Actions } from '../types';
import { REGION, API_VERSION, Environment, SUCCESS_SYMBOL } from '../utils/constants';
import { getGlobalOptions, Command } from '../utils/getGlobalOptions';
import { setAWSCredentials } from '../utils/setAWSCredentials';


interface Input extends Actions {
  name: string;

};

export const deleteParameter = async ({ name, prefix, region = REGION, environment = Environment.all }: Input): Promise<void> => {

  const loader = ora(`deleting key ${name} with prefix /${prefix} (${region})`).start();

  const ssm = new SSM({ apiVersion: API_VERSION, region });

  const keyName = `/${prefix}/${environment}/${name}`;

  const params = {
    Name: keyName,
  }

  try {
    await ssm.deleteParameter(params).promise();
    loader.stopAndPersist({ text: `deleted ${keyName} (${region})`, symbol: SUCCESS_SYMBOL });
  } catch (e) {
    if ((e as AWSError).code === 'ParameterNotFound') {
      loader.stopAndPersist({ text: `parameter ${keyName} not found (${region})`, symbol: SUCCESS_SYMBOL });
    } else {
      loader.fail(`Something went wrong deleting the key ${keyName}, ${e}`);
    }
  }
}


export const command = async (name: string, command: Command): Promise<void> => {

  const { params, credentials } = await getGlobalOptions(command);

  setAWSCredentials(credentials);
  await deleteParameter({ ...params, name });
}