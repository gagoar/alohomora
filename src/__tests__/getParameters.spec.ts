import SSM from '../__mocks__/aws-sdk/clients/ssm';
import { getParameter } from '..';
import { getCommand } from '../actions/commands';
import { stopAndPersist, fail } from '../__mocks__/ora';
import { mockConsole, unMockConsole } from './helpers';

const getParameterPayload = {
  'Name': '/my-company/my-app/production/Vault_713',
  'Type': 'String',
  'Value': "Philosopher's Stone",
  'Version': 1,
  'LastModifiedDate': '2020-04-26T02:03:31.132Z',
  'ARN': 'arn:aws:ssm:us-west-2:687711712713:parameter/my-company/my-app/production/Vault_713',
  'DataType': 'text'
};


describe('getParameters', () => {
  let consoleLogMock: jest.Mock;
  beforeAll(() => {
    consoleLogMock = mockConsole('log');
  });

  afterAll(() => {
    unMockConsole('log');
  });

  beforeEach(() => {
    stopAndPersist.mockReset();
    fail.mockReset();
    consoleLogMock.mockReset();
  });

  it('request fails', async () => {
    const prefix = 'my-company/my-app';

    const handler = jest.fn()
      .mockImplementationOnce(() => { throw new Error('Some fatal error ocurred') });

    SSM.__setResponseForMethods({ getParameter: handler });

    const response = await getParameter({ name: 'Vault_713', environment: 'production', prefix });

    expect(fail).toHaveBeenCalledTimes(1);
    expect(response).toBe('');
  });

  it('gets the parameter', async () => {
    const prefix = 'my-company/my-app';

    const handler = jest.fn(() => ({ Parameter: getParameterPayload }));

    SSM.__setResponseForMethods({ getParameter: handler });

    const response = await getParameter({ name: 'Vault_713', environment: 'production', prefix, ci: true });
    expect(stopAndPersist).toHaveBeenCalledTimes(1);
    expect(response).toMatchSnapshot();
  });

  it('does not find the parameter', async () => {
    const prefix = 'my-company/my-app';

    const error = { code: 'ParameterNotFound', name: 'ParameterNotFound' };

    const handler = jest.fn(() => { throw error; });

    SSM.__setResponseForMethods({ getParameter: handler });

    const response = await getParameter({ name: 'Vault_713', environment: 'production', prefix });
    expect(stopAndPersist).toHaveBeenCalledTimes(1);
    expect(response).toBe('');
  })

  it('does not find the parameter, but it returns Parameter empty, without throwing an Error', async () => {
    const prefix = 'my-company/my-app';

    const handler = jest.fn(() => ({ Parameter: undefined }));

    SSM.__setResponseForMethods({ getParameter: handler });

    const response = await getParameter({ name: 'Vault_713', environment: 'production', prefix });
    expect(stopAndPersist).toHaveBeenCalledTimes(1);
    expect(response).toBe('');
  })

  it('via command invocation', async () => {
    const prefix = 'my-company/my-app';
    const handler = jest.fn(() => ({ Parameter: getParameterPayload }));

    SSM.__setResponseForMethods({ getParameter: handler });

    await getCommand('Vault_713', { parent: { prefix } });

    expect(consoleLogMock).toHaveBeenCalled();
  });
});
