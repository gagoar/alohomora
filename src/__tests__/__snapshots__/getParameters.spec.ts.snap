// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getParameters gets the parameter 1`] = `
"[90m┌───────────[39m[90m┬─────────────────────[39m[90m┬─────────────[39m[90m┬─────────────────────────────────────────────────[39m[90m┬─────────┐[39m
[90m│[39m[31m Name      [39m[90m│[39m[31m Value               [39m[90m│[39m[31m Environment [39m[90m│[39m[31m Updated by                                      [39m[90m│[39m[31m Version [39m[90m│[39m
[90m├───────────[39m[90m┼─────────────────────[39m[90m┼─────────────[39m[90m┼─────────────────────────────────────────────────[39m[90m┼─────────┤[39m
[90m│[39m Vault_713 [90m│[39m Philosopher's Stone [90m│[39m production  [90m│[39m Sunday, April 26th, 2020, 2:03:31 AM (GMT+0000) [90m│[39m 1       [90m│[39m
[90m└───────────[39m[90m┴─────────────────────[39m[90m┴─────────────[39m[90m┴─────────────────────────────────────────────────[39m[90m┴─────────┘[39m"
`;
